mutiny-core/src/federation.rs

```diff
@@ -0,0 +1,612 @@
+
+        let mut client_builder = fedimint_client::Client::builder();
+        client_builder.with_module(WalletClientInit(None));
+        client_builder.with_module(MintClientInit);
+        client_builder.with_module(LightningClientInit);
+
+        let db = g
+            .new_fedimint_client_db(federation_info.federation_id().to_string())
+            .await?
+            .into();
+        if get_config_from_db(&db).await.is_none() {
+            client_builder.with_federation_info(federation_info.clone());
+        }
+
+        client_builder.with_database(db);
+        client_builder.with_primary_module(1);
+
+        let secret = create_federation_secret(xprivkey, federation_index, network)?;
+
+        let fedimint_client = client_builder
+            .build(get_default_client_secret(
+                &secret,
+                &federation_info.federation_id(),
+            ))
+            .await?;
+
+        log_debug!(logger, "Built fedimint client");
+        Ok(FederationClient {
+            uuid,
+            federation_index: federation_index.clone(),
+            federation_code,
+            fedimint_client,
+            stopped_components,
+            storage,
+            network,
+            logger,
+            stop,
+        })
+    }
+
+    pub(crate) fn federation_index(&self) -> FederationIndex {
+        FederationIndex {
+            child_index: self.federation_index.child_index,
+            federation_code: self.federation_code.clone(),
+            archived: self.federation_index.archived,
+        }
+    }
+
+    pub(crate) async fn get_invoice(&self, amount: u64) -> Result<MutinyInvoice, MutinyError> {
+        let lightning_module = self
+            .fedimint_client
+            .get_first_module::<LightningClientModule>();
+        let (_id, invoice) = lightning_module
+            .create_bolt11_invoice(Amount::from_sats(amount), String::new(), None, ())
+            .await?;
+        Ok(invoice.into())
+    }
+
+    /// Get the balance of this federation client in sats
+    pub(crate) async fn get_balance(&self) -> Result<u64, MutinyError> {
+        Ok(self.fedimint_client.get_balance().await.msats / 1_000)
+    }
+
+    pub async fn get_activity(&self) -> Result<Vec<ActivityItem>, MutinyError> {
+        let operations = self
+            .fedimint_client
+            .operation_log()
+            .list_operations(100, None)
+            .await;
+
+        let lightning_module = Arc::new(
+            self.fedimint_client
+                .get_first_module::<LightningClientModule>(),
+        );
+
+        let activity_stream = stream::iter(operations.into_iter());
+
+        let activity_items = activity_stream
+            .then(move |(key, entry)| {
+                let logger = self.logger.clone();
+                let lightning_module = Arc::clone(&lightning_module);
+                async move {
+                    if entry.operation_module_kind() == "ln" {
```

Comment:
is there a way to get around this magic string, can we match on the type instead?